File: BackEnd/src/domain/usecases/weather/GetWeatherForecastUseCase.ts
/**
 * @file GetWeatherForecastUseCase.ts
 * @description Caso de uso para obtener la previsión meteorológica por horas.
 * Delega en WeatherService la lógica de caché + API.
 * @module Weather
 * @layer Domain
 *
 * @implements {IGetWeatherForecastUseCase}
 * @injectable
 * @dependencies WeatherService
 */

import { injectable, inject } from 'inversify';
import { TYPES } from '../../../core/types.js';
import type { IGetWeatherForecastUseCase } from '../../interfaces/usecases/weather/IGetWeatherForecastUseCase.js';
import type { WeatherQueryRequestDTO } from '../../dtos/weather/weather-query-request.dto.js';
import type { ForecastResponseDTO } from '../../dtos/weather/forecast-response.dto.js';
import { WeatherService } from '../../../presentation/services/WeatherService.js';

/**
 * Obtiene la previsión horaria para las coordenadas dadas.
 *
 * @implements {IGetWeatherForecastUseCase}
 * @injectable
 * @dependencies WeatherService
 */
@injectable()
export class GetWeatherForecastUseCase implements IGetWeatherForecastUseCase {
  constructor(
    @inject(TYPES.WeatherService) private readonly weatherService: WeatherService,
  ) {}

  /**
   * @param dto — Localización como string libre (`location`) o coordenadas (`lat`/`lon`),
   *              más número de horas opcionales.
   * @returns ForecastResponseDTO con la lista de condiciones horarias.
   * @throws {ExternalServiceException} Si la API no está disponible.
   */
  async execute(dto: WeatherQueryRequestDTO): Promise<ForecastResponseDTO> {
    if (dto.location?.trim()) {
      return this.weatherService.getForecastForQuery(dto.location.trim(), dto.hours ?? 24);
    }
    return this.weatherService.getForecast(dto.lat!, dto.lon!, dto.hours ?? 24);
  }
}

File: FrontEnd/lib/domain/usecases/weather/get_weather_forecast_use_case.dart
/// @file get_weather_forecast_use_case.dart
/// @description Caso de uso para obtener la previsión meteorológica de N días.
/// Delega en IWeatherRepository la orquestación de caché y llamada al proxy.
/// @module Weather
/// @layer Domain
library;

import '../../entities/weather_data.dart';
import '../../interfaces/usecases/weather/i_get_weather_forecast_use_case.dart';
import '../../repositories/i_weather_repository.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// GET WEATHER FORECAST USE CASE
// ═══════════════════════════════════════════════════════════════════════════════

/// Caso de uso que obtiene la previsión meteorológica para una ubicación.
///
/// Devuelve [WeatherData] con el campo [WeatherData.forecast] poblado.
/// La granularidad de la previsión (días) se controla con el parámetro [days].
///
/// [implements] IGetWeatherForecastUseCase
/// [injectable] registrar en container.dart como lazySingleton.
/// [dependencies] IWeatherRepository.
class GetWeatherForecastUseCase implements IGetWeatherForecastUseCase {
  final IWeatherRepository _repository;

  const GetWeatherForecastUseCase({required IWeatherRepository repository})
      : _repository = repository;

  /// Obtiene la previsión de [days] días para [location].
  ///
  /// [location] — nombre de ciudad o coordenadas aceptadas por WeatherAPI.com.
  /// [days]     — número de días de previsión (1–14, por defecto 7).
  /// [returns]  — [WeatherData] con datos actuales + lista de previsión.
  /// [throws]   AppError.network si no hay conexión y no hay caché válido.
  @override
  Future<WeatherData> execute(String location, {int days = 7}) {
    return _repository.getForecast(location, days: days);
  }
}
